protocol/protocol.go

```diff
@@ -31,194 +46,226 @@ func (m *Manager) routerInitializer(router *routerpkg.Router, netConnection *net
 	outgoingRoute := router.OutgoingRoute()
 
-	m.addFlow("HandleRelayedTransactions", router, []wire.MessageCommand{wire.CmdInv, wire.CmdTx}, isStopping, stop,
-		func(incomingRoute *routerpkg.Route) error {
-			return relaytransactions.HandleRelayedTransactions(m.context, incomingRoute, outgoingRoute)
-		},
-	)
+	return []*flow{
+		m.addFlow("HandleRelayedTransactions", router, []wire.MessageCommand{wire.CmdInv, wire.CmdTx}, isStopping, stop,
+			func(incomingRoute *routerpkg.Route, peer *peerpkg.Peer) error {
+				return relaytransactions.HandleRelayedTransactions(m.context, incomingRoute, outgoingRoute)
+			},
+		)}
 }
 
 func (m *Manager) addFlow(name string, router *routerpkg.Router, messageTypes []wire.MessageCommand, isStopping *uint32,
-	stopChan chan error, flow func(route *routerpkg.Route) error) {
+	stopChan chan error, initializeFunc flowInitializeFunc) *flow {
 
 	route, err := router.AddIncomingRoute(messageTypes)
 	if err != nil {
 		panic(err)
 	}
 
-	spawn(fmt.Sprintf("addFlow-startFlow-%s", name), func() {
-		err := flow(route)
-		if err != nil {
-			m.context.HandleError(err, name, isStopping, stopChan)
-			return
-		}
-	})
+	return &flow{
+		name: name,
+		executeFunc: func(peer *peerpkg.Peer) {
+			err := initializeFunc(route, peer)
+			if err != nil {
+				m.context.HandleError(err, name, isStopping, stopChan)
+				return
+			}
+		},
+	}
 }
 
 func (m *Manager) addOneTimeFlow(name string, router *routerpkg.Router, messageTypes []wire.MessageCommand,
-	isStopping *uint32, stopChan chan error, flow func(route *routerpkg.Route) error) {
+	isStopping *uint32, stopChan chan error, initializeFunc flowInitializeFunc) *flow {
 
 	route, err := router.AddIncomingRoute(messageTypes)
 	if err != nil {
 		panic(err)
 	}
 
-	spawn(fmt.Sprintf("addOneTimeFlow-startFlow-%s", name), func() {
-		defer func() {
-			err := router.RemoveRoute(messageTypes)
+	return &flow{
+		name: name,
+		executeFunc: func(peer *peerpkg.Peer) {
+			defer func() {
+				err := router.RemoveRoute(messageTypes)
+				if err != nil {
+					panic(err)
+				}
+			}()
+
+			err := initializeFunc(route, peer)
 			if err != nil {
-				panic(err)
+				m.context.HandleError(err, name, isStopping, stopChan)
+				return
 			}
-		}()
+		},
+	}
+}
 
-		err := flow(route)
-		if err != nil {
-			m.context.HandleError(err, name, isStopping, stopChan)
-			return
-		}
-	})
+func registerHandshakeRoutes(router *routerpkg.Router) (
+	receiveVersionRoute *routerpkg.Route, sendVersionRoute *routerpkg.Route) {
+	receiveVersionRoute, err := router.AddIncomingRoute([]wire.MessageCommand{wire.CmdVersion})
+	if err != nil {
+		panic(err)
+	}
+
+	sendVersionRoute, err = router.AddIncomingRoute([]wire.MessageCommand{wire.CmdVerAck})
+	if err != nil {
+		panic(err)
+	}
+
+	return receiveVersionRoute, sendVersionRoute
+}
+
+func removeHandshakeRoutes(router *routerpkg.Router) {
+	err := router.RemoveRoute([]wire.MessageCommand{wire.CmdVersion, wire.CmdVerAck})
```

Comment:
This shouldn't work. RemoveRoute should take a Route, not a slice of message types.

Reply:
Talked and agreed that here it doesn't matter. Created ticket.